pkg/api/pod/warnings.go

```diff
@@ -0,0 +1,240 @@
+								Child("matchExpressions").Index(j).
+								Child("key").String()+": "+e.Key+" is "+msg)
+					}
+				}
+			}
+		}
+		for i, t := range n.PreferredDuringSchedulingIgnoredDuringExecution {
+			for j, e := range t.Preference.MatchExpressions {
+				if msg, deprecated := deprecatedNodeLabels[e.Key]; deprecated {
+					warnings = append(
+						warnings,
+						fieldPath.Child("spec", "affinity", "nodeAffinity", "preferredDuringSchedulingIgnoredDuringExecution").Index(i).
+							Child("preference").
+							Child("matchExpressions").Index(j).
+							Child("key").String()+": "+e.Key+" is "+msg)
+				}
+			}
+		}
+	}
+	for i, t := range podSpec.TopologySpreadConstraints {
+		if msg, deprecated := deprecatedNodeLabels[t.TopologyKey]; deprecated {
+			warnings = append(warnings, fieldPath.Child("spec", "topologySpreadConstraints").Index(i).Child("topologyKey").String()+": "+t.TopologyKey+" is "+msg)
+		}
+	}
+
+	// use of deprecated annotations
+	for _, deprecated := range deprecatedAnnotations {
+		if _, exists := meta.Annotations[deprecated.key]; exists {
+			warnings = append(warnings, fieldPath.Child("metadata", "annotations").Key(deprecated.key).String()+": "+deprecated.message)
+		}
+		if len(deprecated.prefix) > 0 {
+			for k := range meta.Annotations {
+				if strings.HasPrefix(k, deprecated.prefix) {
+					warnings = append(warnings, fieldPath.Child("metadata", "annotations").Key(k).String()+": "+deprecated.message)
+					break
+				}
+			}
+		}
+	}
+
+	// removed volume plugins
+	for i, v := range podSpec.Volumes {
+		if v.PhotonPersistentDisk != nil {
+			warnings = append(warnings, fmt.Sprintf("%s: non-functional in v1.16+", fieldPath.Child("spec", "volumes").Index(i).Child("photonPersistentDisk")))
+		}
+	}
+
+	// duplicate hostAliases (#91670, #58477)
+	if len(podSpec.HostAliases) > 1 {
+		items := sets.NewString()
+		for i, item := range podSpec.HostAliases {
+			if items.Has(item.IP) {
+				warnings = append(warnings, fmt.Sprintf("%s: duplicate ip %q", fieldPath.Child("spec", "hostAliases").Index(i).Child("ip"), item.IP))
+			} else {
+				items.Insert(item.IP)
+			}
+		}
+	}
+
+	// duplicate imagePullSecrets (#91629, #58477)
+	if len(podSpec.ImagePullSecrets) > 1 {
+		items := sets.NewString()
+		for i, item := range podSpec.ImagePullSecrets {
+			if items.Has(item.Name) {
+				warnings = append(warnings, fmt.Sprintf("%s: duplicate name %q", fieldPath.Child("spec", "imagePullSecrets").Index(i).Child("name"), item.Name))
+			} else {
+				items.Insert(item.Name)
+			}
+		}
+	}
+	// imagePullSecrets with empty name (#99454#issuecomment-787838112)
+	for i, item := range podSpec.ImagePullSecrets {
+		if len(item.Name) == 0 {
+			warnings = append(warnings, fmt.Sprintf("%s: invalid empty name %q", fieldPath.Child("spec", "imagePullSecrets").Index(i).Child("name"), item.Name))
+		}
+	}
+
+	// duplicate volume names (#78266, #58477)
+	if len(podSpec.Volumes) > 1 {
+		items := sets.NewString()
```

Comment:
This will allocate a bunch.  Not the end of the world, but is someting to keep in mind

Reply:
benchmarks don't show it allocating more than creating a single string set and clearing/reusing it